crates/sui-rest-api/src/transactions/execution.rs

```diff
@@ -356,3 +352,135 @@ fn derive_balance_changes(
         })
         .collect()
 }
+
+pub struct SimulateTransaction;
+
+impl ApiEndpoint<RestService> for SimulateTransaction {
+    fn method(&self) -> axum::http::Method {
+        axum::http::Method::POST
+    }
+
+    fn path(&self) -> &'static str {
+        "/transactions/simulate"
+    }
+
+    fn operation(
+        &self,
+        generator: &mut schemars::gen::SchemaGenerator,
+    ) -> openapiv3::v3_1::Operation {
+        OperationBuilder::new()
+            .tag("Transactions")
+            .operation_id("SimulateTransaction")
+            .query_parameters::<SimulateTransactionQueryParameters>(generator)
+            .request_body(RequestBodyBuilder::new().bcs_content().build())
+            .response(
+                200,
+                ResponseBuilder::new()
+                    .json_content::<TransactionSimulationResponse>(generator)
+                    .bcs_content()
+                    .build(),
+            )
+            .build()
+    }
+
+    fn handler(&self) -> RouteHandler<RestService> {
+        RouteHandler::new(self.method(), simulate_transaction)
+    }
+}
+
+async fn simulate_transaction(
+    State(state): State<Option<Arc<dyn TransactionExecutor>>>,
+    Query(parameters): Query<SimulateTransactionQueryParameters>,
+    accept: AcceptFormat,
+    //TODO allow accepting JSON as well as BCS
+    Bcs(transaction): Bcs<Transaction>,
+) -> Result<ResponseContent<TransactionSimulationResponse>> {
+    let executor = state.ok_or_else(|| anyhow::anyhow!("No Transaction Executor"))?;
+
+    simulate_transaction_impl(&executor, &parameters, transaction).map(|response| match accept {
+        AcceptFormat::Json => ResponseContent::Json(response),
+        AcceptFormat::Bcs => ResponseContent::Bcs(response),
+    })
+}
+
+pub(super) fn simulate_transaction_impl(
+    executor: &Arc<dyn TransactionExecutor>,
+    parameters: &SimulateTransactionQueryParameters,
+    transaction: Transaction,
+) -> Result<TransactionSimulationResponse> {
+    if transaction.gas_payment.objects.is_empty() {
+        return Err(RestError::new(
+            axum::http::StatusCode::BAD_REQUEST,
+            "no gas payment provided",
+        ));
+    }
+
+    let SimulateTransactionResult {
+        input_objects,
+        output_objects,
+        events,
+        effects,
+        mock_gas_id,
+    } = executor
+        .simulate_transaction(transaction.try_into()?)
+        .map_err(anyhow::Error::from)?;
+
+    if mock_gas_id.is_some() {
```

Comment:
Given that there's a previous check on the transaction having a gas object or not, I would expect this `mock_gas_id` to be always None. Why is this check still needed? (I'm probably missing something)

Reply:
That's the point, its just an extra sanity check that nothing unexpected happened.